rs/state_manager/tests/state_manager.rs

```diff
@@ -1223,6 +1223,113 @@ fn missing_manifests_are_recomputed() {
     });
 }
 
+/// Tests that the manifest is computed incrementally using a delta relative to the manifest at a
+/// lower height. Steps are:
+///
+/// - Compute the manifest at height 2 using the delta from the manifest at height 1.
+/// - Compute the manifest at height 3 using the delta from the manifest at height 2.
+/// - Compute the manifests at height 2 and 3 using the deltas from the manifest at height 1.
+///
+/// The third step consists of the first step + a new manifest computation that is expected to
+/// require more hashing than the second step since it's done from height 1.
+///
+/// Asserting that more hashing is required in step 3 ensures two things:
+/// - The computation in the second step was actually done from height 2 since it required less
+///   hashing.
+/// - Incremental manifest computation can be done from a height further back than the previous one
+///   (at the cost of more hashing).
+#[test]
+fn missing_manifest_is_computed_incrementally() {
+    state_manager_restart_test_with_metrics(|_metrics, state_manager, restart_fn| {
+        use ic_state_manager::testing::StateManagerTesting;
+
+        let canister_id = canister_test_id(123);
+
+        let write_stable_memory = |state: &mut ReplicatedState, index: u64| {
+            let mut canister = state.take_canister_state(&canister_id).unwrap();
+            canister
+                .execution_state
+                .as_mut()
+                .unwrap()
+                .stable_memory
+                .page_map
+                .update(&[(PageIndex::new(index), &[1_u8; PAGE_SIZE])]);
+            state.put_canister_state(canister);
+        };
+
+        let purge_manifests_and_restart = |mut state_manager: StateManagerImpl,
+                                           purge_manifest_heights: &[Height],
+                                           restart_height: Height|
+         -> (StateManagerImpl, u64, u64) {
+            for h in purge_manifest_heights {
+                assert!(state_manager.purge_manifest(*h));
+            }
+            let (metrics, state_manager) = restart_fn(state_manager, Some(restart_height));
+            wait_for_checkpoint(&state_manager, restart_height);
+
+            let hashed_key = maplit::btreemap! {"type".to_string() => "hashed".to_string()};
+            let reused_key = maplit::btreemap! {"type".to_string() => "reused".to_string()};
+
+            let chunk_bytes = fetch_int_counter_vec(&metrics, "state_manager_manifest_chunk_bytes");
+
+            // Return the state manager along with the number of reused bytes and the
+            // number of bytes hashed.
+            (
+                state_manager,
+                chunk_bytes[&reused_key],
```

Comment:
```suggestion
                chunk_bytes[&reused_key] + chunk_bytes[&hashed_and_compared_key],
```

The other one could be flaky, as we randomly hash and compare 10% of the chunks.

Reply:
How is `hashed_and_compared` different from just `hashed`? I am basically using two `u64` here. One to make sure the manifest(s) were computed incrementally and the other to determine how much hashing we needed to do to compute a manifest. If I include hashing things into both numbers, I'm not sure the approach still makes sense.

Reply:
Added; as discussed offline.